### PR TITLE
Django Admin Fresh Look

### PR DESCRIPTION
**Pull request by elky**
### Description

Ticket [#24444](https://code.djangoproject.com/ticket/24444#ticket) 

Inspired by a new djangoproject.com site. Looks modern, fresh and clean.

New look contains CSS changes only. Intentionally. All changes are about colors, margins, sizes. Nothing major. No **any** markup changes.

### Requirements
Django 1.8+

### Font
New theme uses [Roboto](http://www.google.com/fonts/specimen/Roboto) webfonts. Font includes Latin, Latin Extended, Cyrilic, Cyrilic Extended, Greek, Greek Extended subsets.

Not sure about others. I will appreciate if someone would test admin interface with **i18n** to find any font issues.

### Testing

Tested in:
- IE7+ (IE7 and IE8 are OK in terms of graceful degradation)
- FF30+ (Windows, Ubuntu, OSX)
- Chrome 35+ (Windows, Ubuntu, OSX)
- Safari 8 (OSX)

#### Login Page
screenshot 2015-03-03 22 28 00

#### Dashboard
screenshot 2015-03-03 22 29 37

#### List of Filtered Objects
screenshot 2015-03-03 22 30 10

#### New Object
screenshot 2015-03-03 22 30 40

### Feedback
Feedback and contribution are welcome.

### CODE REVIEW

**Comment by AtroxDev** [comment]
Wow! That looks great.
:+1:

**Comment by evildmp** [comment]
I think that this needs - perhaps through a single change overriding a default base admin template - to be switchable.

At first, it can be optionally switched in as an admin theme; later, it can become the default.

If it simply replaces the existing theme at any point there will be howls of dismay from people who spent hours colour-matching their own admin tweaks to the blue theme.

**Comment by collinanderson** [comment]
I personally like it. I think the css-only approach is the right way to go. However I have some quick thoughts:

- The deadline for this being included in 1.8 has passed. A pull request against master would be better.
- I feel like there's too much whitespace, especially in the header of the change page. I see about half of what I used to. Is there any way to condense that (I'm mostly thinking vertically)?
image
image
- I assume keeping the gradient in the background of search still was not intentional.
image
- If a change like this is going to succeed where others have failed, it needs to be very conservative. It might make sense to stick to a blueish color palette rather than changing it to green. (and so it doesn't get confused with the django website itself :)
- It might not be possible, but if you can avoid loading an external font, that would be a plus for page load speed.

**Comment by elky** [comment]
@collinanderson 

> The deadline for this being included in 1.8 has passed.

Ah, ok, I missed it. Suggesting to use this PR for discussing - I'll do new one later.

>  Is there any way to condense that (I'm mostly thinking vertically)?

Sure but we have to change markup for this case. And the main goal of this PR is smooth update.
While I did all these changes I played with this action bar (I personally find it annoying to scroll down every time on Change page) and one of the options is to make this bar sticky to the bottom so it always visible and available, see example:
screenshot 2015-03-04 10 51 19

> It might not be possible, but if you can avoid loading an external font, that would be a plus for page load speed.

Agree. I'll just add Roboto font files to the project (it has Apache License 2.0 so we can use it) and load them via font-face.

**Comment by ubernostrum** [comment]
For what it's worth, the font load thing is not just speed -- it also risks exposing to third parties a list of every Django admin instance in the world, and where it's running.

**Comment by sandeepksrana** [comment]
Its realy nice.....

**Comment by elky** [comment]
No external fonts now. Local fonts inclue Latin, Latin Extended, Cyrilic, Cyrilic Extended, Greek, Greek Extended subsets. Not sure about others. I will appreciate if someone would test admin interface with i18n to find any font issues.

**Comment by collinanderson** [comment]
At this point, you could probably get away with just including the woff files. That should work for everything except IE8. I'm worried about the file size of django/contrib/admin/static growing too much. http://caniuse.com/#feat=woff

Also, as Tim mentioned on the ticket, please email https://groups.google.com/d/forum/django-developers for wider feedback. We close pull requests that aren't attached to open tickets or mailing list discussions.

File: django/contrib/admin/static/admin/css/forms.css [hunk 1]
```diff
@@ -158,17 +171,22 @@ fieldset.collapsed .collapse-toggle {
...
+textarea {
+    resize: vertical;
 }
 
 /* SUBMIT ROW */
 
 .submit-row {
-    padding: 5px 7px;
+    padding: 20px;
+    margin: 40px 0 0;
```
Comment (collinanderson) [inline django/contrib/admin/static/admin/css/forms.css]: This 40px margin leads to a lot of extra header whitespace if `save_on_top = True`. Could it be reduced?

Comment (elky) [inline django/contrib/admin/static/admin/css/forms.css]: Ah, I just didn't know about ```save_on_top = True``` feature at all. Yeah, I will fix that.

**Comment by elky** [comment]
@collinanderson 
> At this point, you could probably get away with just including the woff files.

Good point, thanks. By the way, is there any official info about browser support in django admin?

> Please email https://groups.google.com/d/forum/django-developers for wider feedback

I posted a theme in google groups recently but it didn't appear in group still. I think it's not approved yet.

**Comment by collinanderson** [comment]
I don't think there's an official list of supported browsers, but I know 1.9 drops support for IE6&7. https://code.djangoproject.com/ticket/24053

File: django/contrib/admin/static/admin/css/base.css [hunk 2]
```diff
@@ -134,9 +138,9 @@ blockquote {
 }
 
 code, pre {
-    font-family: "Bitstream Vera Sans Mono", Monaco, "Courier New", Courier, monospace;
+    font-family: "Ubuntu Mono", "Bitstream Vera Sans Mono", Monaco, "Courier New", Courier, monospace;
```
Comment (jezdez) [inline django/contrib/admin/static/admin/css/base.css]: Why this change?

**Comment by jezdez** [comment]
I'm not comfortable doing this verbatim port of the website design that is about the Django brand, while the Django admin UI is not.

**Comment by MarkusH** [comment]
I personally like it :+1: but I'd like to see IE8 still supported, at least for one or two more releases.

For the comment @evildmp mentioned above: if we proceed we need some backwards compatibility option. Simply replacing the design from one version to another is not an option for me.

@jezdez raised an important thought: just because the Djangoproject website is now "all-green", this doesn't require the admin to follow that color scheme. (Not being a designer:) what about using the current admin colors with the new styles?

**Comment by jezdez** [comment]
In fact, I'm going to veto the inclusion as I believe this does a disservice to the website design and makes keeping a visual identity for Django hard to separate from the admin UI. I apologize for the time wasted on your part, but I have to close this now. Feel free to bring up the issue on the developer list as suggest before.

**Comment by jacobian** [comment]
As a historical note: the original (green) djangoproject.com theme was deliberately designed to be visually distinct from the (blue) admin theme. Wilson's reasoning, if I can recall it correctly, was that the admin is part of *your* site, not ours, and so should have a visually-distinct theme/brand that can enforce the user's theme, not ours.

Now, that reasoning is literally a decade old (!) so I'm certainly not sure I still agree with it. I'd have to think about it more to figure out where I stand, but that's kinda moot given @jezdez's veto. The blue is certainly looking pretty dated these days, and I'd love to see a visual refresh of the admin, too. 

So I hope we can keep iterating on this, rather then just shut it down entirely.

**Comment by timgraham** [comment]
+1 to continue iterating to distinguish it from the djangoproject.com web site. Would like to see how it looks with the existing admin color scheme as Markus suggested (or perhaps making slight adjustments but keeping the colors neutral rather than going all green.)

**Comment by elky** [comment]
Ok, understood. Thanks for feedback guys. Wait new PR soon :)

**Comment by jezdez** [comment]
Thanks for the background @jacobian, I must say I couldn't agree more with Wilson's reasons, even today.

There are various 3rd party skins (https://www.djangopackages.com/grids/g/admin-interface/) that follow similar ideas as this pull request, modernizing the markup with HTML5 tags, different navigation UIs, even completely new concepts of presenting admin views. All of them have in common that despite design changes they try to stay similarly subtle like Django's current admin skin to make it versatile enough for the flexibility the admin API allows. If you look through the screenshots on those 3rd party packages, you'll see that they often look even more dashboard-y and utilitarian than Django's original skin. I think if we were to redesign it, we should ask ourselves why all the other attempts have gone that direction.

**Comment by aaugustin** [comment]
@elky Other colors would alleviate @jezdez' concerns and a CSS modernization is long overdue.

To make it more likely to get your work accepted, I would suggest to:

- try to find a sponsor within the core team i.e. someone who's sufficiently interested to iterate until your changes look good and then commit them
- not go too far : remember there's a huge installed base and you don't want to create support issues for teams who expose Django admins to users or customers

**Comment by elky** [comment]
@jezdez I personally thought in this way as well to be honest - to make full redesign with lots of changes in markup. But I realized that community will never accept this. Current django admin is old visually and in terms of front-end code, and needs to be refactored. But all changes must be smooth and step by step. That's why I decided to make new look with CSS changes only at first. I was possibly overbold and went too far with green theme, but anyway this attempt was successful - I've got feedback.

@aaugustin thanks for suggestions

**Comment by jezdez** [comment]
@elky Indeed, many thanks for taking that step, I really hope I didn't come over as overprotective. I guess having worked on the redesign implementation recently made me a bit sensitive. Again, please do keep iterating, I'm stoked to see where this leads us.

**Comment by collinanderson** [comment]
@MarkusH The current total size of django/contrib/admin/static is 724K. The woff font files supported by most browsers except IE8 total 327K. The eot font files to support IE8 are 717K on their own. Is it worth it to get the custom font to show up in IE8? I'd prefer the smaller file size and have IE8 gracefully degrade to a built-in font. :)

**Comment by MarkusH** [comment]
@collinanderson Oo, didn't know it was that "bad". Fine with me in that regards then ;)

**Comment by elky** [comment]
@collinanderson just to clarify - are you worrying about static folder size in common or about size of files which are loading? I'm asking because in second case we can find a compromise - how about to use conditional tags tor IE8? This will load extra fonts for IE8 users only.

**Comment by MarkusH** [comment]
@elky Do you mind to summarize the objections that came up here [in your thread on django-developers](https://groups.google.com/forum/#!topic/django-developers/HJAikaEBqJ4) now that your mails got through. That's a more appropriate place for a continued discussion of this topic and it will get a wider audience there as well.

**Comment by collinanderson** [comment]
@elky I'm thinking the size of `manage.py collectstatic`, and to some extent the repo size.

**Comment by thoreg** [comment]
I am wondering how this approach might look like if you take the 'color scheme' of the 'current/old' admin overview. May be this one would get more acceptance.

**Comment by elky** [comment]
@MarkusH thanks, done.

@thoreg @aaugustin @jezdez just made quick example. Is it right way I should follow? (I personally like it)
screenshot 2015-03-05 2 04 08

**Comment by aaugustin** [comment]
Yes, that's interesting. Perhaps not perfectly balanced yet, but I think this could land in 1.9 :-)

**Comment by sbaechler** [comment]
Great styling.
If you used a CSS preprocessor, used variables for the colors and added the source to the repo, users could easily specify their own color palette and just recompile the CSS as part of the build process.
(We use black as the primary color for Django admin on our sites.)

**Comment by elky** [comment]
Sorry for bumping this thread. Just wanted to say I created a blue theme and published an app that easily allows to test it on real project - https://github.com/elky/django-flat-theme.

If you're interested join the discussion here: https://groups.google.com/forum/#!topic/django-developers/HJAikaEBqJ4

Thank you

**Comment by qingfeng** [comment]
:beers:
